File: Advent-of-Code-2021/day03.ts
import * as utils from "./utils.ts";

const lines = await utils.readDay(3);
const numbers: Readonly<number[][]> = lines.map((line) =>
  [...line].map(Number)
);
const BIT_LENGTH = numbers[0].length;
function bitsToInt(bits: number[]): number {
  return parseInt(bits.join(""), 2);
}
const mostCommonValues = numbers
  .map((value) => new utils.Vec(value))
  .reduce((vec, acc) => acc.add(vec), utils.Vec.zeros(BIT_LENGTH))
  .values.map((total) => total >= numbers.length / 2)
  .map(Number);
const leastCommonValues = mostCommonValues.map((v) => 1 - v);
console.log(bitsToInt(mostCommonValues) * bitsToInt(leastCommonValues));

function findRating(keepLeastCommon: boolean) {
  let values = numbers;
  for (const index of utils.range(BIT_LENGTH)) {
    const counts = utils.sum(values.map((number) => number[index]));
    let target = +(counts >= values.length / 2);
    if (keepLeastCommon) {
      target = 1 - target;
    }
    values = values.filter((number) => number[index] === target);
    if (values.length === 1) {
      return values[0];
    }
  }
  throw Error();
}

console.log(bitsToInt(findRating(false)) * bitsToInt(findRating(true)));

File: Advent-of-Code-2021/day03.py
from statistics import mode

numbers = tuple([int(c) for c in s.rstrip()] for s in open("day03.txt").readlines())
BIT_LENGTH = len(numbers[0])
most_common_values = [
    mode(number[index] for number in numbers) for index in range(BIT_LENGTH)
]
least_common_values = [1 - v for v in most_common_values]


def bits_to_int(bits: list[int]) -> int:
    return int("".join(map(str, bits)), 2)


print(bits_to_int(most_common_values) * bits_to_int(least_common_values))


def find_rating(use_least_common: bool) -> list[int]:
    values = numbers
    for index in range(BIT_LENGTH):
        target = sum(v[index] for v in values)
        target = target >= len(values) // 2
        if use_least_common:
            target = 1 - target

        values = [v for v in values if v[index] == target]
        if len(values) == 1:
            return values[0]

    raise Exception()


print(bits_to_int(find_rating(False)) * bits_to_int(find_rating(True)))
